File: kotlin/yacht/src/main/kotlin/Yacht.kt
object Yacht {
    fun solve(category: YachtCategory, vararg dices: Int) = when(category) {
        YachtCategory.ONES -> dices.digitScore(1)
        YachtCategory.TWOS -> dices.digitScore(2)
        YachtCategory.THREES -> dices.digitScore(3)
        YachtCategory.FOURS -> dices.digitScore(4)
        YachtCategory.FIVES -> dices.digitScore(5)
        YachtCategory.SIXES -> dices.digitScore(6)
        YachtCategory.FULL_HOUSE -> dices.fullHouseScore()
        YachtCategory.FOUR_OF_A_KIND -> dices.fourOfAKindScore()
        YachtCategory.LITTLE_STRAIGHT -> dices.littleStraightScore()
        YachtCategory.BIG_STRAIGHT -> dices.bigStraightScore()
        YachtCategory.YACHT -> dices.yachtScore()
        YachtCategory.CHOICE -> dices.choiceScore()
    }

    private fun IntArray.digitScore(digit: Int) = count { it == digit } * digit
    private fun IntArray.fullHouseScore() = diceByCount().let { if (2 in it && 3 in it) sum() else 0 }
    private fun IntArray.fourOfAKindScore() = diceByCount().let { (it[4] ?: it[5] ?: 0) * 4 }
    private fun IntArray.littleStraightScore() = if (straight() && 6 !in this) 30 else 0
    private fun IntArray.bigStraightScore() = if (straight() && 1 !in this) 30 else 0
    private fun IntArray.yachtScore() = if (distinct().size == 1) 50 else 0
    private fun IntArray.choiceScore() = sum()

    private fun IntArray.straight() = distinct().size == 5
    private fun IntArray.diceByCount() = groupBy { it }.map { it.value.size to it.key }.toMap()
}

File: prolog/yacht/yacht.pl
:- use_module(library(clpfd)).

score([X, X, X, X, X], yacht, 50) :- !.

score(Dice, little_straight, 30) :- sort(Dice, [1, 2, 3, 4, 5]), !.
score(Dice, big_straight, 30) :- sort(Dice, [2, 3, 4, 5, 6]), !.

score(Dice, ones, Score)   :- singles_score(Dice, 1, Score), !.
score(Dice, twos, Score)   :- singles_score(Dice, 2, Score), !.
score(Dice, threes, Score) :- singles_score(Dice, 3, Score), !.
score(Dice, fours, Score)  :- singles_score(Dice, 4, Score), !.
score(Dice, fives, Score)  :- singles_score(Dice, 5, Score), !.
score(Dice, sixes, Score)  :- singles_score(Dice, 6, Score), !.

score(Dice, full_house, Score) :-
    (sort(0, @=<, Dice, [X, X, Y, Y, Y]);
     sort(0, @=<, Dice, [X, X, X, Y, Y])),
    X \= Y,
    sum_list(Dice, Score), !.

score(Dice, four_of_a_kind, Score) :-
    (sort(0, @=<, Dice, [_, X, X, X, X]);
     sort(0, @=<, Dice, [X, X, X, X, _])),
    Score #= X * 4, !.

score(Dice, choice, Score) :- sum_list(Dice, Score), !.

score(_, _, 0).

singles_score(Dice, Die, Score) :-
    include(=(Die), Dice, Ones),
    sum_list(Ones, Score).
